### PR TITLE
IPAM persistency

### PR DESCRIPTION
**Pull request by davidefalcone1**
# Description

At the moment, whenever the network manager is (re)started, the IPAM gets info about networks allocated to remote clusters by reading all the tunnelendpoints resource defined in the local cluster.

The startup procedure can be improved by defining a way to persist IPAM configuration, different from the current in-memory structures(since after a reboot, they are lost).

This PR:
1. Defines an interface for the IPAM storage.
2. Implements this interface by means of a storage module based on a new CRD called 'Ipam'

### CODE REVIEW

File: pkg/liqonet/ipamStorage.go [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+func NewIPAMStorage(dynClient dynamic.Interface) (*IPAMStorage, error) {
+	klog.Infof("Init IPAM storage..")
+	ipamStorage := &IPAMStorage{}
+	ipamStorage.dynClient = dynClient
+	klog.Infof("Looking for Ipam resource..")
+	ipam, err := ipamStorage.getConfig()
+	if err != nil && !errors.IsNotFound(err) {
+		return nil, err
+	}
+	if err != nil && errors.IsNotFound(err) {
```
Comment (alacuku) [inline pkg/liqonet/ipamStorage.go:47]: do we need the `err != nil` check?

File: internal/liqonet/tunnelEndpointCreator/tunnelEndpointCreator-config.go [hunk 2]
```diff
@@ -32,60 +27,21 @@ func (tec *TunnelEndpointCreator) SetNetParameters(config *configv1alpha1.Cluste
...
-}
-
 // Helper func that returns a true if the subnet slice passed as first parameter contains the subnet passed as second parameter. Otherwise it returns false
 func (tec *TunnelEndpointCreator) subnetSliceContains(subnetSlice []string, network string) bool {
 	return slice.ContainsString(subnetSlice, network, nil)
 }
 
-// Helper func that remove a subnet from the configuration file
+// Helper func that removes a subnet from the configuration file
 func (tec *TunnelEndpointCreator) removeSubnetFromConfig(network string) []string {
```
[the remark this answers is not in the document]
Comment (davidefalcone1) [reply internal/liqonet/tunnelEndpointCreator/tunnelEndpointCreator-config.go:36]: Actually we don't. Thanks.

File: apis/net/v1alpha1/ipam_types.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// IpamSpec defines the desired state of Ipam
+type IpamSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
```
[the remark this answers is not in the document]
Comment (davidefalcone1) [reply apis/net/v1alpha1/ipam_types.go:30]: No problem

File: apis/net/v1alpha1/ipam_types.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	Prefixes      map[string][]byte `json:"prefixes"`
+	Pools         []string          `json:"pools"`
+	ClusterSubnet map[string]string `json:"clusterSubnet"`
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Cluster
+
+// Ipam is the Schema for the ipams API
+type Ipam struct {
```
Comment (palexster) [inline apis/net/v1alpha1/ipam_types.go:40]: _Ipam_ sounds a bit too generic and prone to confusion for the name of the CRD. What about _IpamStorage_ proposed by @alacuku?

File: pkg/liqonet/ipam.go [hunk 5]
```diff
@@ -188,34 +208,45 @@ func (liqoIPAM *IPAM) GetSubnetPerCluster(network, clusterID string) (string, er
...
+	for _, pool := range ipamConfig.Spec.Pools {
+		klog.Infof("Trying to acquire a child prefix from prefix %s (mask lenght=%d)", pool, getMask(network))
 		if mappedNetwork, err := liqoIPAM.ipam.AcquireChildPrefix(pool, getMask(network)); err == nil {
 			klog.Infof("Network %s has been mapped to network %s", network, mappedNetwork)
-			return mappedNetwork.String(), true
+			return mappedNetwork.String(), nil
 		}
 	}
-	return "", false
+	return "", fmt.Errorf("there are no more networks available")
```
Comment (palexster) [inline pkg/liqonet/ipam.go:249]: For the sake of coherence, the message should start with a capital letter.

File: apis/net/v1alpha1/ipam_types.go (hunk 4 above)
Comment (davidefalcone1) [reply apis/net/v1alpha1/ipam_types.go:40]: IpamStorage sounds good, ok!

File: pkg/liqonet/ipam.go (hunk 5 above)
Comment (davidefalcone1) [reply pkg/liqonet/ipam.go:249]: Golang returns a warning if I try to start an Error with a capital letter. For this reasons, all errors I've ever wrote start with a lowercase letter. Should I change each of them?

File: pkg/liqonet/ipamStorage.go (hunk 1 above)
Comment (davidefalcone1) [reply pkg/liqonet/ipamStorage.go]: No we don't

**Review by palexster — APPROVED** [review]

**Review by alacuku — APPROVED** [review]

**Pull request merged**